src/StackExchange.Redis/ChannelMessageQueue.cs

```diff
@@ -56,64 +63,59 @@ internal ChannelMessage(ChannelMessageQueue queue, RedisChannel channel, RedisVa
     /// <remarks>To create a ChannelMessageQueue, use ISubscriber.Subscribe[Async](RedisKey)</remarks>
     public sealed class ChannelMessageQueue
     {
-        private readonly Channel<ChannelMessage> _queue;
+        private readonly ChannelReader<ChannelMessage> _queue;
+        private readonly Action<CommandFlags> _onUnsubscribe;
+        private readonly Func<CommandFlags, Task> _onUnsubscribeAsync;
+        private readonly Action<Exception> _onInternalError;
+
         /// <summary>
         /// The Channel that was subscribed for this queue
         /// </summary>
         public RedisChannel Channel { get; }
-        private RedisSubscriber _parent;
 
         /// <summary>
         /// See Object.ToString
         /// </summary>
-        public override string ToString() => (string)Channel;
+        public override string ToString() => Channel;
 
         /// <summary>
         /// An awaitable task the indicates completion of the queue (including drain of data)
         /// </summary>
-        public Task Completion => _queue.Reader.Completion;
+        public Task Completion => _queue.Completion;
 
-        internal ChannelMessageQueue(RedisChannel redisChannel, RedisSubscriber parent)
+        /// <summary>
+        /// Constructs a <see cref="ChannelMessageQueue" /> from a <see cref="System.Threading.Channels.ChannelReader{RedisValue}"/> representing
+        /// incoming Redis values on the channel.
+        /// </summary>
+        /// <param name="channel">The name of the channel this subscription is listening on.</param>
+        /// <param name="incomingValues">A channel reader representing the incoming values.</param>
+        /// <param name="onUnsubscribe">A delegate to call when <see cref="Unsubscribe(CommandFlags)"/> is called.</param>
+        /// <param name="onUnsubscribeAsync">A delegate to call when <see cref="UnsubscribeAsync(CommandFlags)"/> is called.</param>
+        /// <param name="onInternalError">REVIEW: Need more context here</param>
```

Comment:
I'm not super clear on the purpose of this `onInternalError` callback (represents the previous call directly to `ConnectionMultiplexer.OnInternalError`). It was only called when a non-`ChannelClosedException` was thrown by `ReadAsync` in the message loop (which is highly unlikely).

Reply:
it was simply to do *something* in the case that an unexpected exception happened - agree "highly unlikely", but: at least in the "real" code, should do *something*; probably doesn't need to be mockable, if it presents a problem

Reply:
We could have two constructors. One that takes a `RedisSubscriber` (i.e. the internal one, that can access `OnInternalError` directly and one that just takes a `ChannelReader` and an action to fire when unsubscribe is called. Since this is primarily (exclusively?) for testing, it's less important that the testable entry point be able to perform all the necessary functions.